flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/RecordWriter.java

```diff
@@ -131,35 +132,30 @@ private void sendToTarget(T record, int targetChannel) throws IOException, Inter
 	}
 
 	public void broadcastEvent(AbstractEvent event) throws IOException, InterruptedException {
-		for (int targetChannel = 0; targetChannel < numChannels; targetChannel++) {
-			RecordSerializer<T> serializer = serializers[targetChannel];
-
-			synchronized (serializer) {
-				Buffer buffer = serializer.getCurrentBuffer();
-				if (buffer != null) {
-					writeAndClearBuffer(buffer, targetChannel, serializer);
-				} else if (serializer.hasData()) {
-					throw new IllegalStateException("No buffer, but serializer has buffered data.");
-				}
-
-				targetPartition.writeEvent(event, targetChannel);
-			}
-		}
-	}
+		final Buffer eventBuffer = EventSerializer.toBuffer(event);
+		try {
+			for (int targetChannel = 0; targetChannel < numChannels; targetChannel++) {
+				RecordSerializer<T> serializer = serializers[targetChannel];
 
-	public void sendEndOfSuperstep() throws IOException, InterruptedException {
-		for (int targetChannel = 0; targetChannel < numChannels; targetChannel++) {
-			RecordSerializer<T> serializer = serializers[targetChannel];
+				synchronized (serializer) {
+					Buffer buffer = serializer.getCurrentBuffer();
+					if (buffer != null) {
+						writeAndClearBuffer(buffer, targetChannel, serializer);
+					} else if (serializer.hasData()) {
+						// sanity check
+						throw new IllegalStateException("No buffer, but serializer has buffered data.");
+					}
 
-			synchronized (serializer) {
-				Buffer buffer = serializer.getCurrentBuffer();
-				if (buffer != null) {
-					writeAndClearBuffer(buffer, targetChannel, serializer);
+					// retain the buffer so that it can be recycled by each channel of targetPartition
+					eventBuffer.retain();
```

Comment:
done, thanks

I opened [FLINK-5277] for the ResultPartition#add test: https://issues.apache.org/jira/browse/FLINK-5277